src/components/advanced-color-picker/advanced-color-picker.test.tsx

```diff
@@ -0,0 +1,309 @@
+        ? `#eee url("data:image/svg+xml,${transparentSvg}")`
+        : value;
+    expect(screen.getByRole("button", { name: "Change colour" })).toHaveStyle({
+      [styleProperty]: styleValue,
+    });
+  }
+);
+
+test("when the close button is clicked, the dialog is closed and the onClose callback function is called", async () => {
+  const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
+
+  const onClose = jest.fn();
+
+  render(<ControlledColorPicker onClose={onClose} />);
+  await user.click(screen.getByRole("button", { name: "Change colour" }));
+  expect(screen.getByRole("dialog")).toBeVisible();
+  expect(onClose).not.toHaveBeenCalled();
+
+  await user.click(screen.getByRole("button", { name: "Close" }));
+  expect(screen.queryByRole("dialog")).not.toBeVisible();
+  expect(onClose).toHaveBeenCalledTimes(1);
+});
+
+test.each([
+  ["enter", "{Enter}"],
+  ["space", " "],
+])(
+  "when the %s key is pressed on a color input with the dialog open, the dialog should close",
+  async (_, keyCode) => {
+    const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
+
+    render(<ControlledColorPicker />);
+
+    await user.click(screen.getByRole("button", { name: "Change colour" }));
+    expect(screen.getByRole("dialog")).toBeVisible();
+
+    screen.getByRole("radio", { name: "pink" }).focus();
+    await user.keyboard(keyCode);
+    jest.runAllTimers();
+
+    expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
+  }
+);
+
+test.each(["a", "b", "q", "t", "x", "4", "0"])(
+  "when the %s key is pressed on a color input with the dialog open, the dialog should not close",
+  async (key) => {
+    const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
+
+    render(<ControlledColorPicker />);
+
+    await user.click(screen.getByRole("button", { name: "Change colour" }));
+    expect(screen.getByRole("dialog")).toBeInTheDocument();
+
+    screen.getByRole("radio", { name: "pink" }).focus();
+    await user.keyboard(key);
+    jest.runAllTimers();
+
+    expect(screen.getByRole("dialog")).toBeVisible();
+  }
+);
+
+test("tabbing from the close button should focus the selected color input", async () => {
+  const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
+
+  render(<ControlledColorPicker selectedColor="#EBAEDE" />);
+  await user.click(screen.getByRole("button", { name: "Change colour" }));
+
+  screen.getByRole("button", { name: "Close" }).focus();
+  await user.keyboard("{Tab}");
+  expect(screen.getByRole("radio", { name: "orchid" })).toBeFocused();
+});
+
+test("tabbing from a color input should focus the close button", async () => {
+  const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
+
+  render(<ControlledColorPicker selectedColor="#EBAEDE" />);
+  await user.click(screen.getByRole("button", { name: "Change colour" }));
+  jest.runAllTimers();
+  expect(screen.getByRole("radio", { name: "orchid" })).toBeFocused();
```

Comment:
`waitFor` is a good suggestion, as you've already suggested elsewhere - I'd forgotten that existed 🤦 

I don't recall seeing any `act` warnings but I may have missed some - I agree that just looking at the code there would seem to be a risk of this.

Reply:
I agree that in some places - not all, but this one is fine - then using `waitFor` is a better solution than `runAllTimers`.

I'm definitely not seeing any `act` warnings for the couple of places I've left `runAllTimers` in - I presume because in these cases no actual React updates are happening (in one case it's to wait for a CSS animation to finish, which is independent of React, in the other it's explicitly to make sure that state updates *don't* happen even when waiting).